internal/config/instance.go

```diff
@@ -123,6 +128,17 @@ func (i *Instance) setWithCallback(key string, valueF func(currentValue interfac
 		return nil
 	}
 
+	pidLock, err := lockfile.NewPidLock(i.lockfile)
+	if err != nil {
+		return errs.Wrap(err, "Could not create lock file for %s", i.lockfile)
+	}
+	defer pidLock.Close()
```

Comment:
This would close the lock after the connection has been created, which doesn't solve the potential conflict. In order for the lock to be valid it would have to be closed when the config connection is closed, but if that were the case then your use-case would not work either as it would encounter the lock..

I think this might be a scenario where it makes more sense to break our rule of "no globals" given the application of the package (api) is so global, and not contextual. Perhaps in time we can properly solve this once and for all by threading a context everywhere.

So I'd suggest calling something like `api.Init(cfg)` from `main.go`. Just make sure that tests are not using this global, because that would break concurrent tests.

Reply:
Maybe I'm misunderstanding how the lockfile library works but I believe it's accessing the lockfile only for database writes and then releasing it. Which should prevent us from running into conflicts between the state tool and the service. I'm realizing now that I'm attempting to solve a problem that already exists so I'll look into the `app.Init(cfg)` approach.